### ISSUE TITLE
LoRA: performance study of the changes from the parent repository

### ISSUE THREAD

**Issue by Andrei-Aksionov** [Lightning-AI/lit-gpt#448]
Hi there 👋 

`TL;DR`: there is a commit[^1] from the parent repo that fixes some bugs and simplifies the code, but my recommendation is to not adopt it as:
1. Those bugs are not relevant to this repo as `fan_in_fan_out` attribute was dropped in one of the latest commits
2. Code simplifications introduces quadratic time complexity with respect to the embeddings size of the model

----

The topic of discussion is the changes[^1] made to the `MergedLinear` class from the parent repository and these changes can be spitted into four parts:
1. A method to merge `lora_A` and `lora_B` matrices.
3. The method above allows to change the order of applying LoRA in forward method: from $x@lora_A@lora_B  \rightarrow x@(lora_A@lora_B)$.
4. The change of the ordering helps to fix the bug: transpose the merged LoRA weights `if fan_in_fan_out==True` before applying them to the input.
5. Since we first merge LoRA weights we can do `zero_pad` on them and is it's just a 2D matrix - the code is much simpler (and more performant!).

Items 1 and 4 are a good addition to the codebase and helps to make the code[^2] much cleaner (just look at the `zero_pad` method!), plus `zero_pad` is at least two times faster[^3] than the current version since there is no need to do transposing two times.
We don't care about Item 3 as `fan_in_fan_out` was dropped in one of the latest commits.

Looks good! 👍  Time to create PR?! Let's first check how much we can win with the new `zero_pad` method. 

I tried a few models on Nvidia T4 and A10G.
Here are the result for A10G:
| Model       | $Loss_{curr}$ | $Loss_{new}$ | $Time_{curr}$ | $Time_{new}$ | $Memory_{curr}$ | $Memory_{new}$ | Precision  |
|-------------|---------------|--------------|---------------|--------------|-----------------|----------------|------------|
| Pythia-70m  | 3.8273        | 3.8693       | 29.22         | 27.25        | 3.04            | 3.00           | bf16-mixed |
| Pythia-410m | 1.8427        | 1.8443       | 90.53         | 87.32        | 8.15            | 7.98           | bf16-mixed |
| Pythia-1B   | 1.7176        | 1.7185       | 115.18        | 130.30       | 13.45           | 13.55          | bf16-mixed |
| Pythia-2.8B | 2.8301        | 2.8271       | 119.95        | 124.66       | 8.54            | 10.26          | bf16-true  |
| StableLM-3B | 3.4154        | 3.4247       | 79.73         | 119.14       | 10.08           | 12.39          | bf16-true  |

$_{curr}$ - current code from the main branch

$_{new}$ - new code from the commit[^1] in the parent repo

As you can see for relatively small models there is a memory and time reduction compared to the code in the main branch, but when the model gets larger the effect is opposite. 
*I also noticed a reduction in memory and time with smaller models on Nvidia T4 and `16-mixed` precision.*

If to take a look at the benchmark notebook[^2] we can see that the `zero_pad` is indeed faster for a very small and relatively big matrices, but the biggest difference is in the code from `forward` method: almost the same time for a very small matrix and more than 10x larger time for a bigger matrices.

What we are missing there? 🤔 
It's time to remember a painful experience from the Leetcode grinding days: enter ~fluggegecheimen~ time complexity analysis!

The biggest difference is that with the current approach in the forward method we first apply `lora_A` to the input `x` to get a matrix of a smaller rank and then apply another smaller rank matrix `lora_B` to it. With the new approach we first apply two smaller rank matrices (`lora_A` and `lora_B`) to each other to get a matrix of a bigger size and then apply it to the input `x`.

Notation:
- $B$: batch size
- $T$: sequence length
- $C_{in}$: number of input features (model dimension)
- $C_{out}$: number of output features ranging from $1\*C_{in}$ to $3\*C_{in}$ (depending on the number of LoRA layers applied to the combined `QKV` matrix)
- $r$: rank
- $e$: number of enabled LoRA layers (to the combined `QKV` matrix)

> [!NOTE] 
I propose to make a couple of simplifications:

1. Calculate time complexity for $e$ of 1 and thus $C_{in} == C_{out} => C$.
2. In addition let's do calculations for a batch size of 1.

## 1. Time complexity for the current approach
1.1) Calculate `after_A` (F.linear): $(B, T, C) @ (r, C).T = (B, T, r)$. We need to do a transpose call and to multiply $C$ numbers $T\*r$ times, so the time complexity is: $C\*T\*r + r\*C$
1.2) Calculate `after_B` (F.conv1d): $(B, T, r).T @ (C, r, 1) = (B, C, T).T = (B, T, C)$. As it's a conv1d we need to multiply $r$ numbers $C$ times (number of filters) for each time-step $T$ and two transpose calls, so the complexity is: $C\*T\*r + T\*r + C\*T$

So the total time complexity to calculate LoRA weights in the forward method is: $C\*T\*r + r\*C + C\*T\*r + T\*r + C\*T=> r\*(2\*C\*T + C + T) + C\*T$

## 2. Time complexity for the new approach
2.1) Calculate `merged_AB` (F.conv1d): $(1, r, C) @ (C, r, 1) = (1, C, C).squeeze = (C, C)$. Time complexity is: $C\*C\*r => C^2\*r$
2.2) Apply `merged_AB` to the input (F.linear): $(B, T, C) @ (C, C).T = (B, T, C)$. Time complexity is: $C\*C\*T + C\*C => C^2\*T + C^2$

So the total time complexity is: $C^2\*(r + T + 1)$

## 3. Conclusion

As you can see with the new approach we have quadratic time complexity with respect to embeddings size which means that the bigger the model the worse the performance.
In other words my recommendation is to not adopt the changes from the commit[^1].   

[^1]: https://github.com/microsoft/LoRA/pull/107
[^2]: https://github.com/Andrei-Aksionov/lit-gpt/blob/00ae8220cb03c8fffec937f5ae2a1450948d421d/lit_gpt/lora.py#L261-L337
[^3]: https://drive.google.com/file/d/1iGgv69N7U4W_RdgpyySPR2HMo2srGCS-/view?usp=sharing

**Comment by carmocca** [comment]
Thank you for the super thorough exploration! I don't have anything to add since everything was already said.

I think we could close this issue now since there's nothing to action on. But it's good that you left this written for posterity

**Comment by Andrei-Aksionov** [comment]
Agree. We can close it.
Didn't know what is the best way to share this knowledge so decided to create an issue.
